### PR TITLE
DRAFT:  uv setup for verl

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

Adds three opt-in extras and a `[tool.uv]` configuration block so the VeOmni
training stack (`actor_rollout_ref.actor.strategy=veomni` + sglang/vllm
rollout) can be materialized via `uv sync` directly out of this repo. This
removes the previous chicken-and-egg where downstream consumers had to
maintain a separate monorepo (e.g. modelchef veomni-recipes) just to express
the curated CUDA 12.9 / Torch 2.9.1 / VeOmni 0.1.9a2 / sglang 0.5.9 pin
matrix.

The existing `pip install -e .` and
`pip install -e .[gpu|vllm|sglang|math|...]` paths are intentionally
unchanged:

- `setup.py`'s `install_requires` and the existing `extras_require` keys
  (`test`, `prime`, `geo`, `gpu`, `math`, `vllm`, `sglang`, `trl`, `mcore`,
  `trtllm`, `transferqueue`) are not touched.
- `pyproject.toml`'s `[project]` block keeps `dynamic = ["version",
  "dependencies", "optional-dependencies", "authors", "urls"]`, so
  setuptools continues to pull metadata from setup.py at build time.
- The new `[tool.uv]` / `[tool.uv.sources]` / `[[tool.uv.index]]` sections
  are uv-only — pip and plain setuptools ignore them entirely.

This PR was prepared with AI assistance; the submitter reviewed every
changed line and ran the smoke tests below.

### Checklist Before Starting

- [x] Search for similar PRs. Queries run:
  - `gh pr list --repo verl-project/verl --state open --search "uv lock"`
  - `gh pr list --repo verl-project/verl --state open --search "pyproject veomni"`
  - `gh pr list --repo verl-project/verl --state open --search "tool.uv"`

  No open PR introduces a uv-managed environment for verl.
- [x] Format the PR title as `[{modules}] {type}: {description}`:
  - `[repo, build, ci] chore: add opt-in uv environment for VeOmni training`
  - Not marking `[BREAKING]`: the new extras (`veomni`, `veomni-vllm`,
    `veomni-sglang`) are purely additive; existing extras keep their
    previous behavior and version pins.

### Test

Reproducible smoke tests on this branch:

```bash
# (1) Lock resolves cleanly. 291 packages across the py3.11/3.12 × Linux
# universe permitted by [tool.uv].environments.
uv lock --python 3.11

# (2) The canonical training combo syncs successfully.
uv sync --python 3.11 --extra veomni --extra veomni-sglang --extra math

# (3) Versions match the modelchef veomni-recipes nightly stack and the
# environment notes in examples/grpo_trainer/run_qwen3_5-35b-a3b_veomni.sh.
.venv/bin/python -c "
import veomni, verl, transformers, sglang, torch, flash_attn, fla
print('veomni', veomni.__version__)              # 0.1.9a2
print('verl', verl.__version__)                  # 0.8.0.dev
print('transformers', transformers.__version__)  # 5.3.0
print('sglang', sglang.__version__)              # 0.5.9
print('torch', torch.__version__)                # 2.9.1+cu129
print('flash_attn', flash_attn.__version__)      # 2.8.4
print('fla', fla.__version__)                    # 0.4.1
"

# (4) The exact symbols verl/workers/engine/veomni/ imports resolve.
.venv/bin/python -c "
from veomni.arguments.arguments_types import OpsImplementationConfig
from veomni.distributed import parallel_state
from veomni.distributed.offloading import build_activation_offloading_context
from veomni.distributed.torch_parallelize import build_parallelize_model
from veomni.models.auto import build_foundation_model
from veomni.ops import apply_ops_config
from veomni.optim import build_lr_scheduler, build_optimizer
from veomni.utils.seqlen_pos_transform_utils import (
    prepare_fa_kwargs_from_position_ids,
)
print('all veomni symbols OK')
"
```

End-to-end validation: the modelchef veomni-recipes nightly suite
(`tests/special_e2e/run_ppo_trainer_veomni.sh`,
`examples/grpo_trainer/run_qwen3_5-35b-a3b_veomni.sh`) targets exactly this
version matrix; once this PR lands, that suite can be re-pointed at this
repo's `.venv` and dropped from the modelchef monorepo.

### API and Usage Example

For pip users — no change:

```bash
# Existing flow keeps working.
pip install -e .
pip install -e ".[gpu]"
pip install -e ".[vllm]"
pip install -e ".[sglang]"
```

For uv users — the new opt-in flow:

#### One-time setup

```bash
# Install uv (the project pins uv==0.9.8 via [tool.uv].required-version,
# so any uv ≥ 0.9.8 works).
curl -LsSf https://astral.sh/uv/install.sh | sh

# Clone and enter the repo (skip if you already have it).
git clone https://github.com/verl-project/verl.git
cd verl
```

#### Sync the curated VeOmni stack

The two recommended combos (mutually exclusive — pick one):

```bash
# (A) sglang rollout — canonical combo, matches the environment notes in
#     examples/grpo_trainer/run_qwen3_5-35b-a3b_veomni.sh
#     (transformers 5.3.0 + sglang 0.5.9 + flashinfer 0.6.3).
uv sync --python 3.11 \
    --extra veomni \
    --extra veomni-sglang \
    --extra math

# (B) vllm rollout — transformers auto-downgrades to 4.57.3 to satisfy
#     vllm 0.14.1's transformers<5 cap.
uv sync --python 3.11 \
    --extra veomni \
    --extra veomni-vllm \
    --extra math
```

What this produces:

- A `.venv/` at the repo root with `verl` installed editable (so edits to
  `verl/workers/engine/veomni/` are picked up immediately, no reinstall).
- The full curated stack: `veomni==0.1.9a2`, `transformers==5.3.0`
  (sglang path) or `4.57.3` (vllm path), `torch==2.9.1+cu129`,
  `flash-attn==2.8.4`, `flash-linear-attention==0.4.1`,
  `liger-kernel`, etc.

#### Activate and verify

```bash
source .venv/bin/activate

python -c "
import veomni, verl, transformers, sglang, torch, flash_attn
print('veomni', veomni.__version__)              # 0.1.9a2
print('verl', verl.__version__)                  # 0.8.0.dev
print('transformers', transformers.__version__)  # 5.3.0
print('sglang', sglang.__version__)              # 0.5.9
print('torch', torch.__version__)                # 2.9.1+cu129
print('flash_attn', flash_attn.__version__)      # 2.8.4
"
```

#### Run a VeOmni training script

```bash
# .venv is active (or use `uv run ...`).
bash examples/grpo_trainer/run_qwen3_5-35b-a3b_veomni.sh
# or, without activating:
uv run --extra veomni --extra veomni-sglang --extra math \
    bash examples/grpo_trainer/run_qwen3_5-35b-a3b_veomni.sh
```

#### Add more extras later

`uv sync` is the single source of truth for the env contents. To add
extras incrementally, re-run `uv sync` with the union of all extras you
want — uv will install / uninstall to converge:

```bash
# Add the geo extra alongside the existing veomni + sglang setup.
uv sync --python 3.11 \
    --extra veomni \
    --extra veomni-sglang \
    --extra math \
    --extra geo
```

Note: re-running `uv sync` *without* an extra will *remove* it from the
env. To keep accumulating, always pass the full `--extra` set, or set
`UV_FROZEN=1` (use existing lock) and use `uv pip install` for ad-hoc
adds.

#### Switch rollouts

`veomni-vllm` and `veomni-sglang` are mutually exclusive (declared in
`[tool.uv].conflicts`). Switching is a single re-sync:

```bash
# Was on sglang; switch to vllm.
uv sync --python 3.11 \
    --extra veomni \
    --extra veomni-vllm \
    --extra math
```

The legacy `vllm` / `sglang` extras are also declared as conflicting with
the new `veomni-*` extras (vllm 0.14.1 vs 0.8–0.12, sglang 0.5.9 vs
0.5.8, plus `quack-kernels` and `grpcio` disagreement between the legacy
sglang pin and the cu129 stack), so combinations like
`--extra veomni --extra vllm` are rejected at lock time with a clear
error.

#### Refresh the lockfile

The committed `uv.lock` is the source of truth. Regenerate (e.g. when
bumping `veomni` or `transformers` pins) with:

```bash
uv lock --python 3.11
git add uv.lock && git commit -m "[repo, build] chore: refresh uv.lock"
```

### Design & Code Changes

Three files changed. Net diff is ~218 lines added, 0 removed (purely
additive).

#### `setup.py` (+53 lines)

Adds three new keys to `extras_require`:

- `VEOMNI_REQUIRES`: `veomni==0.1.9a2` plus the cu129 stack inlined
  (`torch==2.9.1+cu129`, `torchvision==0.24.1+cu129`,
  `torchaudio==2.9.1+cu129`, `torchcodec==0.9.1`, `triton==3.5.1`,
  `liger-kernel`, `flash-attn`, `flash-linear-attention==0.4.1`,
  `cuda-python==12.9.1`, `hf_transfer`).

  We deliberately do *not* use `veomni[gpu]==0.1.9a2` here. veomni's
  `gpu` sub-extra hard-pulls `flash-attn-3`, whose only PyPI release
  (0.0.0) is yanked, and uv applies `[tool.uv.sources]` inconsistently
  for transitive deps. Listing the gpu deps directly lets us pin them via
  `[tool.uv.sources]` / `override-dependencies`.

- `VEOMNI_VLLM_REQUIRES`: `vllm==0.14.1` + `transformers==4.57.3`. The
  4.57.3 pin satisfies vllm 0.14.1's `transformers<5` cap; veomni itself
  doesn't constrain transformers.

- `VEOMNI_SGLANG_REQUIRES`: `sglang==0.5.9` + `transformers==5.3.0` +
  `flashinfer-python==0.6.3`. sglang's metadata pins `transformers==4.57.1`
  but the modelchef nightly stack force-overrides to 5.3.0; the override
  is replicated in `[tool.uv].override-dependencies` below.

#### `pyproject.toml` (+163 lines)

A single new `[tool.uv]` block, plus `[tool.uv.sources]` and
`[[tool.uv.index]]`. Highlights:

- `required-version = "==0.9.8"` — pinned to match the modelchef monorepo
  for reproducible resolves.

- `environments`: restricted to py3.11 / py3.12 on Linux. The prebuilt
  flash-attn / cu129 torch wheels are Linux-x86_64-only, so resolving for
  Win32 / macOS would crash the lock.

- `no-build-isolation-package = ["flash-attn", "flash-attn-3"]`.

- `override-dependencies`: forces the curated cu129 stack (torch+cu129,
  triton 3.5.1, NCCL 2.27.6, cudnn 9.15.1.9, datasets 2.21.0, numpy<2,
  packaging<25.0, diffusers 0.36.0) and per-extra pins for the two
  rollouts (`transformers`, `cuda-python`, `torchcodec`, `opencv`,
  `llguidance`, `outlines-core`, `flashinfer`). The cudnn pin avoids the
  torch 2.9 conv3d perf regression
  (https://github.com/pytorch/pytorch/releases/tag/v2.9.1).

- `conflicts`: forbids the seven impossible combinations (the four
  legacy↔veomni-\* pairs whose pins disagree on `vllm` / `sglang` /
  `grpcio`, the two `veomni` ↔ legacy-vllm/sglang pairs whose pins
  disagree on torch and `quack-kernels`, plus `veomni-vllm` ↔
  `veomni-sglang`). Without these, uv's universal lockfile would fail
  with cross-pin contradictions.

- `dependency-metadata`: short-circuits `pyext==0.7` (pulled by the legacy
  `prime` extra) which fails to build on py3.11+ (uses removed
  `inspect.getargspec`). Declaring zero deps lets `uv lock` succeed
  without building. `uv sync --extra prime` would still fail at install
  time, which matches the upstream reality that `prime` is not py3.11+
  ready.

- `[tool.uv.sources]`: pins the explicit cu129 torch wheel URLs
  (cf. astral-sh/uv#11153) and the cu129 / torch 2.9 prebuilt
  `flash-attn` (per-python wheel) and `flash-attn-3` (cp39-abi3, single
  wheel) wheels. The `extra = "..."` source marker is unavailable here
  because verl's extras are dynamic via setup.py; uv only honors that
  marker for static `[project.optional-dependencies]`. Dropping the
  marker is fine — torch / flash-attn are only introduced by the
  `veomni` extra.

- `[[tool.uv.index]]`: adds the pytorch.org wheel index as `explicit`
  so it is consulted only for the explicitly-mapped torch / torchvision /
  torchaudio packages.

#### `.gitignore` (+2 lines)

Adds `!uv.lock` to override the blanket `*.lock` rule. The committed
`uv.lock` is the source of truth for reproducible installs (this is the
recommended uv pattern, and matches the modelchef monorepo).

Also includes the regenerated `uv.lock` (291 packages, ~5500 lines).

Related upstream:

- VeOmni: ByteDance-Seed/VeOmni, release `v0.1.9a2`.
- uv sources guidance: https://docs.astral.sh/uv/concepts/projects/dependencies/#sources
- cu129 wheel selection bug: astral-sh/uv#11153
- torch 2.9 conv3d regression note:
  https://github.com/pytorch/pytorch/releases/tag/v2.9.1

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: this PR adds a packaging surface only — it does not change Python code paths. End-to-end coverage of the resulting environment runs in modelchef veomni-recipes nightly (Qwen3-MoE / Qwen3.5-MoE GRPO with `strategy: veomni`); reproducing that in upstream verl CI would require pulling VeOmni and a 30B MoE checkpoint, which is out of scope for the existing CI footprint. The lock/sync smoke test in the "Test" section above can be wired into a lightweight CI leg if desired.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by wuxibin89** [comment]
https://github.com/verl-project/verl/pull/5664

**Comment by wuxibin89** [comment]
Close as offline discussion.
